### PR TITLE
bpo-33747: Avoid mutating the global sys.modules dict

### PR DESCRIPTION
**Pull request by onyb**
The test method `test_patch_imports_lazily` was injecting a mocked module
to the `sys.modules` `dict`, which was reused in another test module called
`test_patch_propogrates_exc_on_exit`. Hence, the other test module failed
whenever it was launched independently.

Implemented a context manager to make sure modifications to `sys.modules`
are not persisted after the test methods finish execution.

https://bugs.python.org/issue33747

### CODE REVIEW

File: Lib/unittest/test/testmock/testpatch.py
Comment (encukou) [inline Lib/unittest/test/testmock/testpatch.py]: Could you avoid rearranging the imports? It makes it hard to see what is being changed.

Comment (pablogsal) [inline Lib/unittest/test/testmock/testpatch.py]: Why are these lines changed?

File: Lib/unittest/test/testmock/testpatch.py
Comment (encukou) [inline Lib/unittest/test/testmock/testpatch.py]: What exactly the method does is not very clear from the name. 
There are two places that need to patch in a mocked `squizz` module, but it seems to me that moving this into a method is too much indirection: the testcases would be much easier to read if they included the three lines directly even though it means a little code duplication:

        with swap_attr(sys, 'modules', sys.modules.copy()):
            squizz = Mock()
            sys.modules['squizz'] = squizz
            ...

File: Lib/unittest/test/testmock/testpatch.py
Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: I was rearranging the imports. Just pushed a commit to revert them.

Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: Sorry. I just fixed it in a recent commit.

File: Lib/unittest/test/testmock/testpatch.py
Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: Resolved.

File: Lib/unittest/test/testmock/testpatch.py
Comment (serhiy-storchaka) [inline Lib/unittest/test/testmock/testpatch.py]: I have doubts about changing the reference to `sys.modules`. it can be cached in some places. It is safer to change `sys.modules` in-place. See for example `sys_modules_context` in `Lib/test/test_importlib/test_namespace_pkgs.py`. There are several other helpers in modules test.support and test.test_importlib.util.

Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: Okay with your suggestion, but could you please explain to me what kind of risk is posed by a change in reference?

I like the `sys_modules_context` helper more. Where do you think I should move this function to, before using it in the tests?

Comment (zooba) [reply Lib/unittest/test/testmock/testpatch.py]: I don't know if @brettcannon or any of the other importlib maintainers expect the tests to be standalone, but if not then you could move `sys_modules_context` into `test.support`. Alternatively, it could be copied into `test.support` and leave the importlib tests using their own copy.

Comment (serhiy-storchaka) [reply Lib/unittest/test/testmock/testpatch.py]: > Okay with your suggestion, but could you please explain to me what kind of risk is posed by a change in reference?

I don't remember details (and these details were changed in recent releases), but the comment to `sys_modules_context` implies that there is such risk.

> Where do you think I should move this function to, before using it in the tests?

You can just import it.
```python
from test.test_importlib.test_namespace_pkgs import sys_modules_context
```
There are many import-related helpers scattered through different files, and I think that once we should merge and standardize them, but this is a different issue.

Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: Thanks. I have made the modifications.

**Comment by onyb** [comment]
I have made the requested changes; please review again

File: Lib/unittest/test/testmock/testpatch.py
Comment (brettcannon) [inline Lib/unittest/test/testmock/testpatch.py]: I know that @serhiy-storchaka recommended using this context manager, but since it isn't in `test.test_importlib.util` then I wouldn't consider it stable. Instead, I would use [`test.test_importlib.util.uncache()`](https://github.com/python/cpython/blob/077059e0f086cf8c8b7fb9d1f053e38ddc743f59/Lib/test/test_importlib/util.py#L121) which will do what you need and is considered usable by other tests.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Minor swap to another context manager which is supported for use by other test modules.

File: Lib/unittest/test/testmock/testpatch.py
Comment (serhiy-storchaka) [reply Lib/unittest/test/testmock/testpatch.py]: `test.test_importlib.util.uncache()` is less strong than `test.test_importlib.test_namespace_pkgs.sys_modules_context()`, but I think in this case it would be enough.

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/unittest/test/testmock/testpatch.py
Comment (onyb) [reply Lib/unittest/test/testmock/testpatch.py]: Resolved.

**Comment by onyb** [comment]
I have made the requested changes; please review again

**Review by brettcannon — APPROVED** [review]

**Pull request merged**
